File: frontend/lib/features/gamification/domain/entities/group_achievement.dart
import 'package:equatable/equatable.dart';

enum GroupAchievementType {
  groupTotalHabits,
  groupAllStreak,
  groupPerfectWeek;

  String get displayName {
    switch (this) {
      case GroupAchievementType.groupTotalHabits:
        return 'Совместные привычки';
      case GroupAchievementType.groupAllStreak:
        return 'Общая серия';
      case GroupAchievementType.groupPerfectWeek:
        return 'Идеальная неделя команды';
    }
  }
}

class GroupAchievement extends Equatable {
  const GroupAchievement({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.type,
    required this.conditionValue,
    required this.rewardPoints,
    required this.isEarned,
    required this.progress,
    required this.progressPercent,
    this.earnedAt,
  });

  final String id;
  final String name;
  final String description;
  final String icon;
  final GroupAchievementType type;
  final int conditionValue;
  final int rewardPoints;
  final bool isEarned;
  final int progress;
  final double progressPercent;
  final DateTime? earnedAt;

  bool get isInProgress => !isEarned && progress > 0;
  bool get isLocked => !isEarned && progress == 0;

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        icon,
        type,
        conditionValue,
        rewardPoints,
        isEarned,
        progress,
        progressPercent,
        earnedAt,
      ];
}

File: backend/app/domain/models/group_achievement.py
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from uuid import UUID


class GroupAchievementType(str, Enum):
    GROUP_TOTAL_HABITS = "group_total_habits"
    GROUP_ALL_STREAK = "group_all_streak"
    GROUP_PERFECT_WEEK = "group_perfect_week"


@dataclass
class GroupAchievement:
    id: UUID
    name: str
    description: str
    icon: str
    achievement_type: GroupAchievementType
    condition_value: int
    reward_points: int
    is_active: bool

    def check_condition(self, group_value: int) -> bool:
        return group_value >= self.condition_value


@dataclass
class EarnedGroupAchievement:
    id: UUID
    group_id: UUID
    achievement_id: UUID
    earned_at: datetime
    notified: bool

    def mark_as_notified(self) -> None:
        self.notified = True
